File: frontend/react-web/src/features/notifications/services/systemNotificationDelivery.ts
/**
 * System Notification Delivery — Web Notifications API integration.
 *
 * Delivers native OS notifications using the Web Notifications API.
 * Designed to be called from the notification check cycle (Web Worker or main thread).
 *
 * Uses a static label mapping (not React i18n) because the Web Worker
 * cannot access React context. Locale is determined by the document language
 * or a passed locale parameter.
 *
 * **Validates: Requirements 2.3, 2.4, 2.5, 2.8, 5.4, 5.5**
 */

/** Maximum length for notification title (event name) */
const MAX_TITLE_LENGTH = 65;

/** Path to the Planixor 192×192 app icon used in system notifications */
const NOTIFICATION_ICON_PATH = '/icons/icon-192x192.png';

/**
 * Time remaining label translations indexed by alert offset and locale.
 *
 * Covers the four supported offsets:
 * - 0: "Now" / "Ahora"
 * - 10: "In 10 minutes" / "En 10 minutos"
 * - 60: "In 1 hour" / "En 1 hora"
 * - 1440: "In 1 day" / "En 1 día"
 */
const TIME_REMAINING_LABELS: Record<number, Record<string, string>> = {
  0: { en: 'Now', es: 'Ahora' },
  10: { en: 'In 10 minutes', es: 'En 10 minutos' },
  60: { en: 'In 1 hour', es: 'En 1 hora' },
  1440: { en: 'In 1 day', es: 'En 1 día' },
};

/**
 * Truncates an event name to the maximum allowed title length.
 * If the name exceeds 65 characters, it is trimmed and no ellipsis is added
 * (the OS notification system handles overflow display).
 */
export const truncateTitle = (eventName: string): string => {
  if (eventName.length <= MAX_TITLE_LENGTH) {
    return eventName;
  }
  return eventName.slice(0, MAX_TITLE_LENGTH);
};

/**
 * Formats a date and time into a localized display string.
 * Example: "20 jun 2025 · 10:00" (Spanish) or "Jun 20, 2025 · 10:00" (English)
 */
export const formatDateTime = (
  startDay: string,
  startTime: number,
  locale: string,
): string => {
  const parts = startDay.split('-');
  const date = new Date(Number(parts[0]), Number(parts[1]) - 1, Number(parts[2]));
  const hours = Math.floor(startTime / 60);
  const minutes = startTime % 60;
  const timeStr = `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}`;

  const dateStr = date.toLocaleDateString(
    locale.toLowerCase().startsWith('es') ? 'es-ES' : 'en-US',
    {
      day: 'numeric',
      month: 'short',
      year: 'numeric',
    },
  );

  return `${dateStr} · ${timeStr}`;
};

/**
 * Formats the time remaining label for a given alert offset and locale.
 * Falls back to a generic "In X minutes" / "En X minutos" format for unknown offsets.
 */
export const formatTimeRemaining = (
  alertOffset: number,
  locale: string,
): string => {
  const labels = TIME_REMAINING_LABELS[alertOffset];
  if (labels) {
    const lang = locale.toLowerCase().startsWith('es') ? 'es' : 'en';
    return labels[lang] ?? labels['en'] ?? '';
  }
  const lang = locale.toLowerCase().startsWith('es') ? 'es' : 'en';
  return lang === 'es'
    ? `En ${alertOffset} minutos`
    : `In ${alertOffset} minutes`;
};

/**
 * Detects the current document locale.
 * Returns the lang attribute of the document, falling back to 'en'.
 *
 * In a Web Worker context where `document` is not available,
 * this function returns 'en' as the default.
 */
export const detectLocale = (): string => {
  if (typeof document !== 'undefined' && document.documentElement?.lang) {
    return document.documentElement.lang;
  }
  return 'en';
};

/**
 * Delivers a system notification via the Web Notifications API.
 *
 * Checks:
 * 1. `Notification` global is available (not in all Web Worker contexts)
 * 2. Permission is "granted"
 *
 * If either check fails, returns false — the notification record should
 * retain `isDelivered=false` and be reattempted on the next check cycle.
 *
 * @param eventIcon - The emoji icon for the event (from shift or reminder)
 * @param eventName - The calendar event name (will be truncated to 65 chars)
 * @param startDay - ISO date string "YYYY-MM-DD" for the event start day
 * @param startTime - Minutes from midnight for the event start time
 * @param alertOffset - The alert offset in minutes (0, 10, 60, or 1440)
 * @param locale - Optional locale override. If not provided, auto-detects from document.
 * @returns true if the notification was successfully created, false otherwise
 */
export const deliverSystemNotification = (
  eventIcon: string,
  eventName: string,
  startDay: string,
  startTime: number,
  alertOffset: number,
  locale?: string,
): boolean => {
  // Check 1: Web Notifications API is available
  if (typeof Notification === 'undefined') {
    return false;
  }

  // Check 2: Permission is granted
  if (Notification.permission !== 'granted') {
    return false;
  }

  const resolvedLocale = locale ?? detectLocale();
  const title = `${eventIcon} ${truncateTitle(eventName)}`;

  const dateTimeLine = formatDateTime(startDay, startTime, resolvedLocale);
  const remainingLine = formatTimeRemaining(alertOffset, resolvedLocale);
  const body = `${dateTimeLine}\n${remainingLine}`;

  new Notification(title, {
    body,
    icon: NOTIFICATION_ICON_PATH,
  });

  return true;
};

File: frontend/android-app/app/src/main/java/com/codenized/planixor/data/notification/SystemNotificationDelivery.kt
package com.codenized.planixor.data.notification

import android.content.Context
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.codenized.planixor.PlanixorApplication.Companion.NOTIFICATION_CHANNEL_ID
import com.codenized.planixor.R
import dagger.hilt.android.qualifiers.ApplicationContext
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Handles delivery of system notifications via [NotificationManagerCompat].
 * Checks whether notifications are enabled before posting.
 * Returns a Boolean indicating delivery success/failure for routing logic.
 */
@Singleton
class SystemNotificationDelivery @Inject constructor(
    @ApplicationContext private val context: Context,
) {

    /**
     * Delivers a system notification for a due notification record.
     *
     * @param notificationId Unique integer ID for the notification (used for updates/cancellation).
     * @param eventIcon The emoji icon for the event (from shift or reminder).
     * @param eventName The calendar event name displayed in the notification title (truncated to 65 chars).
     * @param startDay ISO date string "YYYY-MM-DD" for the event start day.
     * @param startTime Minutes from midnight for the event start time.
     * @param alertOffset The alert offset in minutes before event start.
     * @return true if the notification was posted successfully, false if notifications are disabled.
     */
    fun deliver(
        notificationId: Int,
        eventIcon: String,
        eventName: String,
        startDay: String,
        startTime: Int,
        alertOffset: Int,
    ): Boolean {
        val notificationManager = NotificationManagerCompat.from(context)

        if (!notificationManager.areNotificationsEnabled()) {
            return false
        }

        val title = "$eventIcon ${eventName.take(MAX_EVENT_NAME_LENGTH)}"

        val dateTimeLine = formatDateTime(startDay, startTime)
        val remainingLine = formatTimeRemaining(alertOffset)
        val body = "$dateTimeLine\n$remainingLine"

        val notification = NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_launcher_foreground)
            .setContentTitle(title)
            .setContentText(dateTimeLine)
            .setStyle(NotificationCompat.BigTextStyle().bigText(body))
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setAutoCancel(true)
            .build()

        try {
            notificationManager.notify(notificationId, notification)
        } catch (e: SecurityException) {
            return false
        }

        return true
    }

    /**
     * Formats the date and time line for the notification body.
     * Example: "20 jun 2025 · 10:00"
     */
    private fun formatDateTime(startDay: String, startTime: Int): String {
        val date = LocalDate.parse(startDay)
        val hours = startTime / 60
        val minutes = startTime % 60
        val timeStr = String.format(Locale.ROOT, "%02d:%02d", hours, minutes)
        val dateFormatter = DateTimeFormatter.ofPattern("d MMM yyyy", Locale.getDefault())
        val dateStr = date.format(dateFormatter)
        return "$dateStr · $timeStr"
    }

    /**
     * Formats the time remaining line based on the alert offset.
     */
    private fun formatTimeRemaining(alertOffset: Int): String {
        return when {
            alertOffset == 0 -> context.getString(R.string.notification_remaining_now)
            alertOffset < 60 -> context.getString(R.string.notification_remaining_minutes, alertOffset)
            alertOffset == 60 -> context.getString(R.string.notification_remaining_1_hour)
            alertOffset == 1440 -> context.getString(R.string.notification_remaining_1_day)
            else -> context.getString(R.string.notification_remaining_minutes, alertOffset)
        }
    }

    companion object {
        private const val MAX_EVENT_NAME_LENGTH = 65
    }
}
